src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableMessagingHelpers.cs

```diff
@@ -2,11 +2,215 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
 using System.Runtime;
+using System.ServiceModel.Security;
+using System.Threading;
 using System.Threading.Tasks;
+using System.Xml;
 
 namespace System.ServiceModel.Channels
 {
+    internal delegate Task OperationWithTimeoutAsyncCallback(TimeSpan timeout);
+
+    internal sealed class Guard
+    {
+        private TaskCompletionSource<object> _tcs;
+        private int _currentCount = 0;
+        private int _maxCount;
+        private bool _closed;
+        private object _thisLock = new object();
+
+        public Guard() : this(1) { }
+
+        public Guard(int maxCount)
+        {
+            _maxCount = maxCount;
+        }
+
+        public void Abort()
+        {
+            _closed = true;
+        }
+
+        public async Task CloseAsync(TimeSpan timeout)
+        {
+            lock (_thisLock)
+            {
+                if (_closed)
+                    return;
+
+                _closed = true;
+
+                if (_currentCount > 0)
+                    _tcs = new TaskCompletionSource<object>();
+            }
+
+            if (_tcs != null)
+            {
+                try
+                {
+                    if (!await _tcs.Task.AwaitWithTimeout(timeout))
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new TimeoutException(SR.Format(SR.TimeoutOnOperation, timeout)));
+                }
+                finally
+                {
+                    lock (_thisLock)
+                    {
+                        _tcs.TrySetResult(null);
+                        _tcs = null;
+                    }
+                }
+            }
+        }
+
+        public bool Enter()
+        {
+            lock (_thisLock)
+            {
+                if (_closed)
+                    return false;
+
+                if (_currentCount == _maxCount)
+                    return false;
+
+                _currentCount++;
+                return true;
+            }
+        }
+
+        public void Exit()
+        {
+            lock (_thisLock)
+            {
+                _currentCount--;
+
+                if (_currentCount < 0)
+                {
+                    throw Fx.AssertAndThrow("Exit can only be called after Enter.");
+                }
+
+                if (_currentCount == 0)
+                {
+                    if (_tcs != null)
+                        _tcs.TrySetResult(null);
```

Comment:
Consider using SetResult instead of TrySetResult. SetResult will fail if the TCS is already completed, which probably indicates an error in your code. TrySetResult will just do nothing if the TCS is already completed.

Reply:
Validating logic and preconditions is more of an assertion check. I'll add an assert which throws on a debug build to make sure the task hasn't already been completed. TrySetResult is a marginally cheaper method.